python_ta/reporters/json_reporter.py

```diff
@@ -33,8 +33,36 @@ def display_messages(self, layout: BaseLayout) -> None:
             output.append(
                 {
                     "filename": k,
-                    "msgs": [msg.to_dict() for msg in msgs],
+                    "msgs": self._output_messages(msgs),
                 }
             )
 
         self.writeln(json.dumps(output, indent=4))
+
+    def _output_messages(self, msgs: List[NewMessage]) -> List[Dict]:
+        """Returns a list of dictionaries containing formatted error messages."""
+        max_messages = self.linter.config.pyta_number_of_messages
+        output_lst = []
+        all_msg_names = [msg.message.msg_id for msg in msgs]
+
+        unique_msg_names = []
+        [
+            unique_msg_names.append(msg_id)
+            for msg_id in all_msg_names
+            if msg_id not in unique_msg_names
+        ]
+
+        for msg_id in unique_msg_names:
```

Comment:
This code is okay but is less efficient than it could be. Instead try this approach:

- loop over just `msgs`
- use a `dict` accumulator mapping error `msg_id` to the count of that message type "seen so far" in the loop. Note that you can initialize the accumulator using all of the unique keys (see above comprehension) with initial corresponding value 0
- then inside the loop use the accumulator to decide whether to append a message to `output_lst` or not
- after that loop, use a second loop over `output_lst` that sets the `number_of_occurrences` for each output message---we don't need to bother with updating this for messages that aren't included in the output!